src/check/constrain/unify/ty.rs

```diff
@@ -1,39 +1,42 @@
-use EitherOrBoth::Both;
-use itertools::{EitherOrBoth, Itertools};
+use std::collections::HashMap;
 
 use crate::check::constrain::constraint::Constraint;
-use crate::check::constrain::constraint::expected::Expect::{Collection, Tuple, Type};
-use crate::check::constrain::constraint::expected::Expected;
+use crate::check::constrain::constraint::expected::{Expect, Expected};
+use crate::check::constrain::constraint::expected::Expect::Type;
 use crate::check::constrain::constraint::iterator::Constraints;
 use crate::check::constrain::Unified;
-use crate::check::constrain::unify::expression::substitute::substitute;
+use crate::check::constrain::unify::expression::SubRes;
 use crate::check::constrain::unify::finished::Finished;
 use crate::check::constrain::unify::link::unify_link;
 use crate::check::context::{Context, LookupClass};
-use crate::check::name::{Any, ColType, IsSuperSet, Name};
-use crate::check::name::name_variant::NameVariant;
+use crate::check::name::{Any, ContainsTemp, IsSuperSet, Name, NameMap, Substitute};
 use crate::check::result::{TypeErr, TypeResult};
 use crate::common::position::Position;
 use crate::common::result::WithCause;
 
-pub fn unify_type(
-    constraint: &Constraint,
-    constraints: &mut Constraints,
-    finished: &mut Finished,
-    ctx: &Context,
-    total: usize,
-) -> Unified {
+pub fn unify_type(constraint: &Constraint, constr: &mut Constraints, finished: &mut Finished,
+                  ctx: &Context, total: usize) -> Unified {
     let (left, right) = (&constraint.parent, &constraint.child);
-    let count = if constraints.len() <= total { total - constraints.len() } else { 0 };
+    let count = if constr.len() <= total { total - constr.len() } else { 0 };
 
     match (&left.expect, &right.expect) {
         (Type { name: l_ty }, Type { name: r_ty }) => {
             if l_ty.is_temporary() {
-                substitute_ty(right.pos, r_ty, left.pos, l_ty, constraints, count, total)?;
-                return unify_link(constraints, finished, ctx, total);
+                substitute_ty(right.pos, r_ty, left.pos, l_ty, constr, count, total)?;
+                return unify_link(constr, finished, ctx, total);
             } else if r_ty.is_temporary() {
-                substitute_ty(left.pos, l_ty, right.pos, r_ty, constraints, count, total)?;
-                return unify_link(constraints, finished, ctx, total);
+                substitute_ty(left.pos, l_ty, right.pos, r_ty, constr, count, total)?;
+                return unify_link(constr, finished, ctx, total);
+            } else if l_ty.contains_temp() {
+                for (old, new) in l_ty.temp_map(r_ty, left.pos)? {
```

Comment:
The last two blocks in this chain I think supersede the top two.
That is, `temp_map` should also return a map if self is temporary.
And, `contains_temp` is also true if self is temporary.

Reply:
Hm apparently not.